src/clients/FtpGateway.cc

```diff
@@ -1042,9 +1042,9 @@ Ftp::Gateway::checkAuth(const HttpHeader * req_hdr)
     /* Test URL login syntax. Overrides any headers received. */
     loginParser(request->url.userInfo(), true);
 
-    /* name is missing. that's fatal. */
-    if (!user[0])
-        fatal("FTP login parsing destroyed username info");
+    /* empty user is not an error; if someone is intent on passing that to
+     * us, go ahead and pass it to the server; this will almost certainly
+     * fail but at least fail gracefully */
```

Comment:
>  I think we could keep it consistent with the rest of the code by simply adding a `|| strlen(login.c_str()) == 0` check to the `Ftp::Gateway::loginParser` to prevent empty usernames from overwriting anything that came before.

Please use a `!login[0]` test instead:

```C++
if (login.isEmpty())
    return;

if (!login[0]) {
    debugs(9, 2, "WARNING: Ignoring FTP credentials that start with a NUL character");
    // TODO: Either support credentials with NUL characters (in any position) or ban all of them.
    return;
}
```

I would also add `Assure(user[0])` after rfc1738_unescape() but that is optional. AFAICT, that terrible function cannot make `user` empty in this context.

Finally, I would replace the fatal() `if` statement with:

```suggestion
    // XXX: We we keep default "anonymous" instead of properly supporting empty usernames.
    Assure(user[0]);
```

Reply:
I changed the commit according to your requests. Looks good to me (log is from a backport of the relevant change to v4.16) 
`printf "GET ftp://ftp.dlptest.com/ HTTP/1.1\r\nAuthorization: Basic AG00\r\n\r\n" | nc <squid_host> <squid_proxy_port>`
causes (excerpt)
```
2023/11/03 12:43:40.571 kid1| 9,2| src/clients/FtpClient.cc(810) writeCommand: ftp<< USER anonymous^M  
2023/11/03 12:43:40.642 kid1| 9,2| src/clients/FtpClient.cc(1127) parseControlReply: ftp>> 331 Please specify the password.   
2023/11/03 12:43:40.642 kid1| 9,2| src/clients/FtpClient.cc(810) writeCommand: ftp<< PASS anonymous@foo^M 
2023/11/03 12:43:43.571 kid1| 9,2| src/clients/FtpClient.cc(1127) parseControlReply: ftp>> 530 Login incorrect. 
...
2023/11/03 12:43:43.577 kid1| 11,2| src/http/Stream.cc(272) sendStartOfMessage: HTTP Client REPLY:                                                                                                                                       
---------                                                                                                                                                                                                                                                     
HTTP/1.1 401 Unauthorized^M 
...
X-Squid-Error: ERR_FTP_FORBIDDEN 530^M     
```

Reply:
> Hope I did not mess up the whole review process by force-pushing

You did not, but there is usually no reason to force-push. In the vast majority of cases, it only makes things more difficult for humans working on the PR. You also do not need to merge master into your PR branch (unless there are conflicts).